spectrum-offchain-cardano/src/tx_submission.rs

```diff
@@ -140,39 +138,148 @@ where
+            }
         }
-    }
-}
 
-impl Display for TxRejected {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        match &self {
-            TxRejected::MissingInputs(inputs) => write!(f, "TxRejected::MissingInputs({:?})", inputs),
-            TxRejected::Unknown => write!(f, "TxRejected::Unknown"),
+        if !outputs.is_empty() {
+            Ok(outputs)
+        } else {
+            Err("No missing inputs")
         }
     }
 }
 
 #[async_trait::async_trait]
-impl<const ERA: u16, Tx> Network<Tx, TxRejected> for TxSubmissionChannel<ERA, Tx>
+impl<const ERA: u16, Tx> Network<Tx, RejectReasons> for TxSubmissionChannel<ERA, Tx>
 where
     Tx: Send,
 {
-    async fn submit_tx(&mut self, tx: Tx) -> Result<(), TxRejected> {
+    async fn submit_tx(&mut self, tx: Tx) -> Result<(), RejectReasons> {
         let (snd, recv) = oneshot::channel();
         self.0.send(SubmitTx(tx, snd)).await.unwrap();
         recv.await.expect("Channel closed").into()
     }
 }
+
+#[derive(Debug, Clone, derive_more::Display, derive_more::From)]
+#[display(fmt = "RejectReasons: {:?}", "_0")]
+pub struct RejectReasons(pub Vec<ProtocolRejectReason>);
+
+#[derive(Debug, Clone, derive_more::Display)]
+pub enum ProtocolRejectReason {
+    /// Missing/bad TX inputs
+    #[display(fmt = "ProtocolRejectReason::BadInputsUtxo: {:?}", "_0")]
+    BadInputsUtxo(Vec<TxInput>),
+    #[display(
+        fmt = "ProtocolRejectReason::ValueNotConserved: Consumed {:?}, produced {:?}",
+        consumed,
+        produced
+    )]
+    ValueNotConserved { consumed: Value, produced: Value },
+    #[display(fmt = "ProtocolRejectReason::MissingVKeyWitnesses: {:?}", "_0")]
+    MissingVKeyWitnesses(Vec<pallas_crypto::hash::Hash<28>>),
+    #[display(fmt = "ProtocolRejectReason::MissingScriptWitnesses: {:?}", "_0")]
+    MissingScriptWitnesses(Vec<pallas_crypto::hash::Hash<28>>),
+    /// A generic error that cannot be clearly identified.
+    Unknown(String),
+}
+
+impl From<ShelleyLedgerPredFailure> for ProtocolRejectReason {
+    fn from(value: ShelleyLedgerPredFailure) -> Self {
+        match value {
+            ShelleyLedgerPredFailure::UtxowFailure(babbage_failure) => match babbage_failure {
+                BabbageUtxowPredFailure::AlonzoInBabbageUtxowPredFailure(alonzo_fail) => match alonzo_fail {
+                    AlonzoUtxowPredFailure::ShelleyInAlonzoUtxowPredfailure(shelley_fail) => {
+                        match shelley_fail {
+                            ShelleyUtxowPredFailure::MissingVKeyWitnessesUTXOW(m) => {
+                                ProtocolRejectReason::MissingVKeyWitnesses(m)
+                            }
+                            ShelleyUtxowPredFailure::MissingScriptWitnessesUTXOW(m) => {
+                                ProtocolRejectReason::MissingScriptWitnesses(m)
+                            }
+                            ShelleyUtxowPredFailure::InvalidWitnessesUTXOW => unimplemented!(),
```

Comment:
`unimplemented!()` is a bad idea, bot should tolerate errors

Reply:
The code here will never panic because the decoding in the Pallas PR safely skips over all unimplemented decodings. How about we convert all these variants into `ProtocolRejectReason::Unknown(_)`?